sql/src/main/java/io/crate/breaker/CircuitBreakerRamAccounting.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.breaker;
+
+import org.elasticsearch.common.breaker.CircuitBreaker;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * RamAccounting implementation which uses a CircuitBreaker to short circuit operations
+ * which would take up too much memory.
+ */
+@ThreadSafe
+public final class CircuitBreakerRamAccounting implements RamAccounting {
```

Comment:
Hm the `RamAccountingContext` also uses a `CircuitBreaker`. maybe better name it something like `DirectRamAccounting` or find a better name for `RamAccountingContext`? 
Maybe `EagerlyBlockBasedRamAccounting`, `FlushingBlockBasedRamAcounting` and `DirectRamAccounting`..

Reply:
Not sure about `DirectRamAccounting`, "direct memory" is often used to refer to off-heap memory, so it might be a bit confusing.

My plan was to try to get rid of the `RamAccountingContext`, that's why I didn't bother too much about the conflicting naming of `CircuitBreakerRamAccounting`. But might be a while to get there.

Reply:
An option might be to de-couple the `CircuitBreakerRamAccounting` further - into a `ConcurrentRamAccounting` that also just uses a `IntConsumer` to account & de-account

Reply:
I like this option of `ConcurrentRamAccounting`, but no strong opinion here. Also ok to leave it like that for now and head further to remove the `RamAccountingContext` as you said.